packages/dev-server-hmr/src/HmrPlugin.ts

```diff
@@ -0,0 +1,282 @@
+    if (context.path === NAME_HMR_CLIENT_IMPORT) {
+      if (!this._webSockets) {
+        return;
+      }
+      return hmrClientScript(this._webSockets);
+    }
+
+    // We are serving a new file or it has changed, so clear all the
+    // dependencies we previously tracked (if any).
+    this._clearDependencies(context.path);
+    this._logger?.debug(`[hmr] Cleared dependency tree cache of ${context.path}`);
+  }
+
+  /** @inheritDoc */
+  async transformImport({ source, context }: { source: string; context: Context }) {
+    const importPath = pathUtil.resolve(context.path, source);
+
+    if (importPath === NAME_HMR_CLIENT_IMPORT) {
+      return;
+    }
+
+    const mod = this._getOrCreateModule(context.path);
+    const dependencyMod = this._getOrCreateModule(importPath);
+
+    mod.dependencies.add(importPath);
+    dependencyMod.dependents.add(context.path);
+    this._logger?.debug(`[hmr] Added dependency from ${context.path} -> ${importPath}`);
+  }
+
+  /** @inheritDoc */
+  async transform(context: Context) {
+    // If the module references import.meta.hot it can be assumed it
+    // supports hot reloading
+    const hmrEnabled = context.body.includes('import.meta.hot') === true;
+    const mod = this._getOrCreateModule(context.path);
+    mod.hmrEnabled = hmrEnabled;
+    this._logger?.debug(`[hmr] Setting hmrEnabled=${hmrEnabled} for ${context.path}`);
+
+    if (hmrEnabled && context.response.is('js')) {
+      return `
+        import {create as __WDS_HMR__} from '${NAME_HMR_CLIENT_IMPORT}';
+        import.meta.hot = __WDS_HMR__(import.meta.url);
+        ${context.body}
+      `;
+    }
+  }
+
+  /**
+   * Clears the dependency cache/tree of a particular module
+   * @param path Module path to clear
+   */
+  protected _clearDependencies(path: string): void {
+    const mod = this._getModule(path);
+
+    if (!mod) {
+      return;
+    }
+
+    for (const dep of mod.dependencies) {
+      const depMod = this._getModule(dep);
+      if (depMod) {
+        depMod.dependents.delete(path);
+      }
+    }
+
+    mod.dependencies = new Set();
+  }
+
+  /** @inheritDoc */
+  serverStop() {
+    this._webSockets = undefined;
+  }
+
+  /**
+   * Fired when a file has changed.
+   * @param path Module path which has changed
+   */
+  protected _onFileChanged(path: string): void {
+    this._triggerUpdate(path);
+  }
+
+  /**
+   * Triggers an update for a given module.
+   * This will result in the client being sent a message to tell them
+   * how to deal with this module updating.
+   * @param path Module path to update
+   * @param visited Modules already updated (cache)
+   */
+  protected _triggerUpdate(path: string, visited: Set<string> = new Set()): void {
+    // We already visited this module
+    if (visited.has(path)) {
+      return;
+    }
+
+    const mod = this._getModule(path);
+    visited.add(path);
+
+    // We have never encountered this module, so can't do anything.
+    if (!mod) {
```

Comment:
at the min i assume if we see an unknown file change, it can't possibly be in the current page as the parents of said file would've also changed (and triggered updates anyway). and we don't want to reload because you could reload a page which has nothing to do with this file. so i do nothing...

if anyone can think of a reason to not do nothing, do let me know

Reply:
I think that's fine to start